File: lytecache-php/src/Support/Paths.php
<?php

declare(strict_types=1);

namespace Lytecache\Support;

/**
 * Default database path resolution, shared by LyteCache and its static
 * defaultPath() accessor.
 */
final class Paths
{
    private const ENV_OVERRIDE = 'LYTECACHE_PATH';

    /**
     * Resolves the default database file location:
     * "<platform cache dir>/lytecache/<project-id>.db", or the
     * LYTECACHE_PATH environment variable if set (after "~" expansion,
     * but not otherwise forced to an absolute path -- a relative
     * LYTECACHE_PATH stays relative, matching the other implementations).
     */
    public static function defaultPath(): string
    {
        $override = getenv(self::ENV_OVERRIDE);
        if ($override !== false && $override !== '') {
            return self::expandHome($override);
        }

        $cacheDir = self::platformCacheDir();
        $projectId = self::projectId();

        return $cacheDir.DIRECTORY_SEPARATOR.'lytecache'.DIRECTORY_SEPARATOR.$projectId.'.db';
    }

    private static function platformCacheDir(): string
    {
        return match (PHP_OS_FAMILY) {
            'Darwin' => self::homeDir().'/Library/Caches',
            'Windows' => self::windowsLocalAppData(),
            default => self::xdgCacheHome(),
        };
    }

    private static function windowsLocalAppData(): string
    {
        $localAppData = getenv('LOCALAPPDATA');
        if ($localAppData !== false && $localAppData !== '') {
            return $localAppData;
        }

        return self::homeDir().'\\AppData\\Local';
    }

    private static function xdgCacheHome(): string
    {
        $xdg = getenv('XDG_CACHE_HOME');
        if ($xdg !== false && $xdg !== '') {
            return $xdg;
        }

        return self::homeDir().'/.cache';
    }

    private static function homeDir(): string
    {
        $home = getenv('HOME');
        if ($home !== false && $home !== '') {
            return $home;
        }

        // Windows fallback when HOME isn't set -- USERPROFILE is the native variable there.
        $userProfile = getenv('USERPROFILE');
        if ($userProfile !== false && $userProfile !== '') {
            return $userProfile;
        }

        throw new \RuntimeException('lytecache: could not determine the home directory to derive the default cache path');
    }

    /**
     * The first 12 hex characters of the SHA-256 hash of the resolved,
     * absolute current working directory -- the same derivation used by
     * the Python, Java, Node.js, and Go implementations of lytecache, so
     * a process in any of those languages started from the same
     * directory resolves to the same file.
     */
    public static function projectId(): string
    {
        $cwd = getcwd();
        if ($cwd === false) {
            $cwd = '.';
        }

        $resolved = realpath($cwd);
        if ($resolved === false) {
            $resolved = $cwd;
        }

        return substr(hash('sha256', $resolved), 0, 12);
    }

    /** Expands a leading "~" or "~/" to the user's home directory. */
    public static function expandHome(string $path): string
    {
        if ($path === '~') {
            return self::homeDir();
        }

        if (str_starts_with($path, '~/') || str_starts_with($path, '~'.DIRECTORY_SEPARATOR)) {
            return self::homeDir().DIRECTORY_SEPARATOR.substr($path, 2);
        }

        return $path;
    }
}

File: lytecache-go/paths.go
package lytecache

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// envPathOverride is checked before the platform default; see DefaultPath.
const envPathOverride = "LYTECACHE_PATH"

// DefaultPath returns the default database file location for the current
// working directory: "<platform cache dir>/lytecache/<project-id>.db".
//
// The platform cache dir comes from [os.UserCacheDir], which already
// implements the XDG Base Directory spec on Linux, ~/Library/Caches on
// macOS, and %LocalAppData% on Windows.
//
// <project-id> is the first 12 hex characters of the SHA-256 hash of the
// resolved, absolute current working directory -- the same derivation used
// by the Python, Java, and Node.js implementations of lytecache, so a
// process in any of those languages started from the same directory
// resolves to the same file.
//
// If the LYTECACHE_PATH environment variable is set (after "~" expansion),
// it is returned instead, taking priority over the derived default. A
// relative LYTECACHE_PATH is returned as-is, not resolved to absolute --
// matching the other implementations.
func DefaultPath() (string, error) {
	if override, ok := os.LookupEnv(envPathOverride); ok && override != "" {
		return expandHome(override)
	}

	cacheDir, err := os.UserCacheDir()
	if err != nil {
		return "", fmt.Errorf("lytecache: resolving platform cache dir: %w", err)
	}

	id, err := projectID()
	if err != nil {
		return "", err
	}

	return filepath.Join(cacheDir, "lytecache", id+".db"), nil
}

// projectID derives the 12-hex-character project identifier from the
// resolved current working directory. See DefaultPath's doc comment for
// the cross-language contract this must uphold.
func projectID() (string, error) {
	cwd, err := os.Getwd()
	if err != nil {
		return "", fmt.Errorf("lytecache: resolving working directory: %w", err)
	}

	resolved, err := filepath.EvalSymlinks(cwd)
	if err != nil {
		// A cwd that can't be resolved (e.g. removed out from under the
		// process) still needs a deterministic id, so fall back to the
		// unresolved path rather than failing outright.
		resolved = cwd
	}

	sum := sha256.Sum256([]byte(resolved))
	return hex.EncodeToString(sum[:])[:12], nil
}

// expandHome expands a leading "~" or "~/" to the user's home directory.
// It does not otherwise touch the path -- a relative path stays relative.
func expandHome(path string) (string, error) {
	if path != "~" && !strings.HasPrefix(path, "~/") {
		return path, nil
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("lytecache: expanding ~ in %s: %w", envPathOverride, err)
	}
	if path == "~" {
		return home, nil
	}
	return filepath.Join(home, path[2:]), nil
}
